pkg/kapp/diff/change_set_with_versioned_rs.go

```diff
@@ -76,6 +84,76 @@ func (d ChangeSetWithVersionedRs) Calculate() ([]Change, error) {
 	return allChanges, nil
 }
 
+func (d ChangeSetWithVersionedRs) checkAndUpdateResWithMaxDurationAnn(existingRsGrouped map[string][]ctlres.Resource,
+	existingRs versionedResources) error {
+
+	for _, res := range d.newRs {
+		if val, found := res.Annotations()[maxDurationAnnKey]; found {
+			duration, err := time.ParseDuration(val)
+			if err != nil {
+				return fmt.Errorf("Unable to parse max-duration for resource %s: %s", res.Description(), err.Error())
+			}
+
+			existingNonVerRsGrouped := d.groupNonVerResources(existingRs.NonVersioned)
+
+			resKey := VersionedResource{res, nil}.UniqVersionedKey().String()
+			exRes, found := existingNonVerRsGrouped[resKey]
+			if !found {
+				if exResGroup, found := existingRsGrouped[resKey]; found {
+					// exRes contains list of all versioned resources currently present on the cluster for particular resource and are sorted
+					// here we will be using last created resource.
+					exRes = exResGroup[len(exResGroup)-1]
+				}
+			}
+
+			if exRes != nil {
+				var lastRenewed time.Time
+				val := exRes.Annotations()[lastRenewTimeAnnKey]
+				if val != "" {
+					lastRenewed, err = time.Parse(time.RFC3339, val)
+					if err != nil {
+						return fmt.Errorf("Unable to parse last-renewed-time for resource %s: %s", res.Description(), err.Error())
+					}
+				}
+
+				err = updateLastRenewedTime(res, duration, lastRenewed, exRes.CreatedAt())
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func updateLastRenewedTime(res ctlres.Resource, duration time.Duration, lastRenewed, creationTime time.Time) error {
+
+	if creationTime.Before(lastRenewed) {
```

Comment:
`creationTime` will be always before `lastRenewed` in case of subsequent update scenario. Am I missing something?

Reply:
If resource got recreated for some reason(reason could be ->user has made some update before max-duration expiry) after we updated last-renewed-time then it will be. This can happen if user has set update-strategy as always-replace for non-versioned resource or resource is versioned.